TEPHRA-171 Added ZkClientService in Tephra which does not extend from…

sagarkapare opened the pull request:
… guava Service.

JIRA: https://issues.cask.co/browse/TEPHRA-171

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/TransactionAdmin.java:
```diff
@@ -29,7 +29,6 @@
 import com.google.inject.Guice;
 import com.google.inject.Injector;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.twill.zookeeper.ZKClientService;
```
Removed all occurrences of Twill's `ZKClientService` from tephra.

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/ZKClientService.java:
```diff
@@ -0,0 +1,118 @@
+  /**
+   * Initiates service shutdown (if necessary), returning once the service has finished stopping.
+   */
+  void stopAndWait();
```
`ZKClientService` interface now has `startAndWait` and `stopAndWait` methods.

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/runtime/DiscoveryModules.java:
```diff
@@ -82,9 +82,7 @@ private DiscoveryService providesDiscoveryService(final ZKClientService zkClient
       return new DiscoveryService() {
         @Override
         public Cancellable register(Discoverable discoverable) {
-          if (!zkClient.isRunning()) {
```
zkClient no longer supports `isRunning()` method now.

chtyim replied:
Why not have the `isRunning()` method in `ZKClientService`?

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/runtime/ZKModule.java:
```diff
@@ -52,16 +62,68 @@ private ZKClientService provideZKClientService(Configuration conf) {
       zkStr = conf.get(TxConstants.HBase.ZOOKEEPER_QUORUM);
     }
 
-    return ZKClientServices.delegate(
```
Not using `ZKClientServices.delegate` since it returns `org.apache.twill.zookeeper.ZKClientService` instance.

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/runtime/ZKModule.java:
```diff
@@ -52,16 +62,68 @@ private ZKClientService provideZKClientService(Configuration conf) {
+                                                 RetryStrategies.exponentialDelay(500, 2000, TimeUnit.MILLISECONDS));
+
+    final ZKClient client = ZKClients.reWatchOnExpire(zkClient);
+    return new ForwardingZKClientService(zkClientService) {
```
Copied `ForwardingZKClientService` to Tephra which now extends from Tephra's `ZKCLientService`.

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/zookeeper/BasicACLData.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * A straightforward implementation of {@link ACLData}.
+ */
+final class BasicACLData implements ACLData {
```
Copied `BasicACLData`, `BasicNodeChildren`, and `BasicNodeData` from Twill since they were in `org.apache.twill.internal.zookeeper` package on which Tephra does not depend.

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/zookeeper/DefaultZKClientService.java:
```diff
@@ -0,0 +1,639 @@
+  private final List<Watcher> connectionWatchers;
+  private final Multimap<String, byte[]> authInfos;
+  private final AtomicReference<ZooKeeper> zooKeeper;
+  private final Service serviceDelegate;
```
Note that `serviceDelegate` is still the guava Service instance.

sagarkapare commented on tephra-core/src/main/java/co/cask/tephra/zookeeper/DefaultZKClientService.java:
```diff
@@ -0,0 +1,627 @@
+
+  @Override
+  public void startAndWait() {
+    ((ServiceDelegate) serviceDelegate).connectToZookeeper();
```
Explicitly calling `connectToZookeeper` to synchronously connecting to the zookeeper. This may not be the best method since we require casting here.

sagarkapare commented:
@chtyim @poornachandra Please take a look. Thanks!

chtyim commented on tephra-core/src/main/java/co/cask/tephra/ZKClientService.java:
```diff
@@ -0,0 +1,118 @@
+  /**
+   * Initiates service startup (if necessary), returning once the service has finished starting.
+   */
+  void startAndWait();
```
Just call it `start` and `stop`.

chtyim commented on tephra-core/src/main/java/co/cask/tephra/zookeeper/BasicNodeChildren.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.List;
+
+/**
+ *
```
Add a line of javadoc.

chtyim commented:
Just one comment. Otherwise the code change LGTM. Also make sure we don't use `ZKClientService.Builder` in the rest of the code base.

sagarkapare commented:
Thank you @chtyim for the review. Made sure we are not using `ZKClientService.Builder` in tephra. Added javadoc and also removed the unused constructor from the `TephraZKClientService`. Will merge once the build passes.

sagarkapare commented:
Build passed. Merging.
